eng/New-TestResources.ps1

```diff
@@ -0,0 +1,305 @@
+}
+if ($AdditionalParameters) {
+    $templateParameters += $AdditionalParameters
+}
+
+# Try to detect the shell based on the parent process name (e.g. launch via shebang).
+$shell, $shellExportFormat = if (($parentProcessName = (Get-Process -Id $PID).Parent.ProcessName) -and $parentProcessName -eq 'cmd') {
+    'cmd', 'set {0}={1}'
+} elseif (@('bash', 'csh', 'tcsh', 'zsh') -contains $parentProcessName) {
+    'shell', 'export {0}={1}'
+} else {
+    'PowerShell', '$env:{0} = ''{1}'''
+}
+
+foreach ($templateFile in $templateFiles) {
+    # Deployment fails if we pass in more parameters than are defined.
+    Write-Verbose "Removing unnecessary parameters from template '$templateFile'"
+    $templateJson = Get-Content -LiteralPath $templateFile | ConvertFrom-Json
+    $templateParameterNames = $templateJson.parameters.PSObject.Properties.Name
+
+    $templateFileParameters = $templateParameters.Clone()
+    foreach ($key in $templateParameters.Keys) {
+        if ($templateParameterNames -notcontains $key) {
+            Write-Verbose "Removing unnecessary parameter '$key'"
+            $templateFileParameters.Remove($key)
+        }
+    }
+
+    $preDeploymentScript = $templateFile | Split-Path | Join-Path -ChildPath 'test-resources-pre-deploy.ps1'
+    if (Test-Path $preDeploymentScript) {
+        Log "Invoking pre-deployment script '$preDeploymentScript'"
+        &$preDeploymentScript -ResourceGroupName $resourceGroupName @PSBoundParameters
+    }
+
+    Log "Deploying template '$templateFile' to resource group '$($resourceGroup.ResourceGroupName)'"
+    $deployment = New-AzResourceGroupDeployment -Name $BaseName -ResourceGroupName $resourceGroup.ResourceGroupName -TemplateFile $templateFile -TemplateParameterObject $templateFileParameters
+    if ($deployment.ProvisioningState -eq 'Succeeded') {
+        # New-AzResourceGroupDeployment would've written an error and stopped the pipeline by default anyway.
+        Write-Verbose "Successfully deployed template '$templateFile' to resource group '$($resourceGroup.ResourceGroupName)'"
+    }
+
+    if ($deployment.Outputs.Count -and !$env:SYSTEM_TEAMPROJECTID) {
+        # Write an extra new line to isolate the environment variables for easy reading.
+        Log "Persist the following environment variables based on your detected shell ($shell):`n"
+    }
+
+    $deploymentOutputs = @{}
+    foreach ($key in $deployment.Outputs.Keys) {
+        $variable = $deployment.Outputs[$key]
+
+        # Work around bug that makes the first few characters of environment variables be lowercase.
+        $key = $key.ToUpperInvariant()
+
+        if ($variable.Type -eq 'String' -or $variable.Type -eq 'SecureString') {
+            $deploymentOutputs[$key] = $variable.Value
+
+            if ($env:SYSTEM_TEAMPROJECTID) {
+                # Running in Azure Pipelines.
+                Write-Host "##vso[task.setvariable variable=$key;issecret=true;]$($variable.Value)"

```

Comment:
It's slightly unfortunate that all variables would become secrets, having some visible might benefit debugging. We can fix this later.

Reply:
There's no good way to differentiate, unfortunately. We have to change output SecureStrings to Strings to even get the value, so everything coming to us here would be String anyway (in practice) which, I agree, would be good to switch on for the value of `issecret`.

I will add a comment, though, explaining this / why and maybe we can change it in the future if we can think of a better way.